include/swift/Runtime/Metadata.h

```diff
@@ -2125,6 +2125,99 @@ struct GenericWitnessTable {
   void *PrivateData[swift::NumGenericMetadataPrivateDataWords];
 };
 
+/// The structure of a type metadata record.
+///
+/// This contains enough static information to recover type metadata from a
+/// name. It is only emitted for types that do not have an explicit protocol
+/// conformance record. 
+///
+/// This structure is notionally a subtype of a protocol conformance record
+/// but as we cannot change the conformance record layout we have to make do
+/// with some duplicated code.
+struct TypeMetadataRecord {
+private:
+  // Some description of the type that is resolvable at runtime.
+  union {
+    /// A direct reference to the metadata.
+    RelativeIndirectablePointer<Metadata> DirectType;
+    
+    /// An indirect reference to the metadata.
+    RelativeIndirectablePointer<const ClassMetadata *> IndirectClass;
```

Comment:
We'd only ever emit type metadata records for types in the current object file, so the indirectability isn't necessary. There should also be records for generic metadata patterns.